designs/extensible-cli-and-scaffolding-plugins-phase-2.md

```diff
@@ -0,0 +1,326 @@
+// written by the plugin following a certain command.
+type PluginResponse struct {
+  // Command holds the command that gets executed by the plugin such as init, create api, etc.
+  Command       string                   `json:"command"`
+
+  // Help contains the plugin specific help text that the plugin returns to kubebuilder when it receives
+  // `--help` flag from Kubebuilder.
+  Help          string                   `json:"help,omitempty"`
+
+  // APIVersion defines the versioned schema of the PluginResponse that will be written back to kubebuilder.
+  // Initially, this will be marked as alpha (v1alpha1).
+  APIVersion    string                   `json:"apiVersion"`
+
+  // Universe in the PluginResponse represents the updated file contents that was written by the plugin.
+  Universe      map[string]string        `json:"universe"`
+
+  // Error is a boolean type that indicates whether there were any errors due to plugin failures.
+  Error         bool                     `json:"error,omitempty"`
+
+  // ErrorMsg holds the specific error message of plugin failures.
+  ErrorMsg      string                   `json:"error_msg,omitempty"`
+}
+```
+
+These structs would be defined on the Kubebuilder side. `PluginRequest` holds all the information `kubebuilder` receives from the CLI and the request will be marshaled into a JSON and sent over `stdin`. `PluginResponse` is what the plugin constructs with the updated universe and sent back to `kubebuilder`.
+
+```go
+func (p *ExternalPlugin) runExternalProgram(req PluginRequest) (res PluginResponse, err error) {
+  pluginReq, err := json.Marshal(req)
+  if err != nil {
+    return res, err
+  }
+
+  cmd := exec.Command(p.Path)
+  cmd.Dir = p.DirContext
+  cmd.Stdin = bytes.NewBuffer(pluginReq)
+  cmd.Stderr = os.Stderr
+
+  out, err := cmd.Output()
+  if err != nil {
+    fmt.Fprint(os.Stdout, string(out))
+    return res, err
+  }
+
+  if json.Unmarshal(out, &res); err != nil {
+    return res, err
+  }
+  return res, nil
+}
+```
+
+The above function handles construction of the `PluginRequest` based on the information `kubebuilder` receives from the CLI and the request is marshaled into JSON. The command to run the external plugin by providing the plugin path will be invoked and `kubebuilder` will send the marshaled `PluginRequest` JSON to the plugin over `stdin`.
+
+On the plugin side, the request JSON will be decoded and depending on what the `Command` in the `PluginRequest` is, the corresponding function to handle `init` or `create api` will be invoked thereby modifying the universe by writing the updated files to it. After `init` or `create api` functions execute successfully, the plugin will write back `PluginResponse` with updated universe and errors (if any) in JSON format through `stdout` to `kubebuilder`. `PluginResponse` also contains error fields `Error` and `ErrorMsg` that the plugin can utilize to add error context if any errors occur.
```

Comment:
Do we have a JSON schema or sample to show what it will look like to the plugins? If I'm writing a plugin in Python I would need to know what the JSON will look like so I can properly handle it on my side.

Reply:
I added example JSON for both `PluginRequest and `PluginResponse` in the Implementation details section. Let me know if that would suffice.

Reply:
@rashmigottipati yes, that is good enough for me.